### PR TITLE
Enable arbitrary euler angle for Mag rotation [NEW]

### PR DESCRIPTION
**Pull request by bresch**
### Solved Problem
replaces https://github.com/PX4/PX4-Autopilot/pull/20247

### Changelog Entry
For release notes:
```
Allow arbitrary rotation of external mag
New parameter: CAL_MAG[0-3]_[ROLL|PITCH|YAW]
Documentation: need update
```

### Test coverage
- Bench tests

### CODE REVIEW

File: src/lib/conversion/rotation.h [hunk 1]
```diff
@@ -91,8 +91,10 @@ enum Rotation : uint8_t {
 	ROTATION_ROLL_90_PITCH_68_YAW_293 = 38,
 	ROTATION_PITCH_315                = 39,
 	ROTATION_ROLL_90_PITCH_315        = 40,
+	ROTATION_MAX,
 
-	ROTATION_MAX
+	// Rotation Enum reserved for custom rotation using Euler Angles
+	ROTATION_CUSTOM                  = 100
```
Comment (dagar) [inline src/lib/conversion/rotation.h:97]: Nevermind, I didn't realize `100` already made it into mavlink. [MAV_SENSOR_ROTATION_CUSTOM](http://mavlink.io/en/messages/common.html#MAV_SENSOR_ROTATION_CUSTOM)

File: src/lib/sensor_calibration/Magnetometer.cpp [hunk 2]
```diff
@@ -201,12 +217,42 @@ bool Magnetometer::ParametersLoad()
...
+			    && ((fabsf(euler_roll_deg) > FLT_EPSILON)
+				|| (fabsf(euler_pitch_deg) > FLT_EPSILON)
+				|| (fabsf(euler_yaw_deg) > FLT_EPSILON))) {
+				rotation_value = ROTATION_CUSTOM;
+				SetCalibrationParam(SensorString(), "ROT", _calibration_index, rotation_value);
+				/* EVENT
+				 * @description <param>CAL_MAG{1}_ROT</param> is set to CUSTOM because
+				 * <param>CAL_MAG{1}_ROLL</param>, <param>CAL_MAG{1}_PITCH</param> and <param>CAL_MAG{1}_YAW</param> are set.
+				 */
+				events::send<int8_t>(events::ID("mag_rot_forced_custom"), events::Log::Info,
```
Comment (dagar) [inline src/lib/sensor_calibration/Magnetometer.cpp:239]: I'm not sure this is worth an event.

**Comment by dagar** [comment]
You need to save the ROLL/PITCH//YAW parameters here.

https://github.com/PX4/PX4-Autopilot/blob/a42400dcd575348eb0f7529b57b57b4c6a3624fa/src/lib/sensor_calibration/Magnetometer.cpp#L353-L358

**Comment by RomanBapst** [comment]
@dagar @bresch I was helping @KonradRudin to figure out why the MAVROS rover tests are failing on his PR. Looks like it fails here as well.
From what I could see the mean error between the ground truth yaw and the estimated yaw is too large.
I plotted it and it looks like this:
image

Any ideas?

**Comment by bresch** [comment]
added cross-ref of the params to find them more easily:
Screenshot from 2023-08-25 16-07-47

**Comment by bresch** [comment]
@RomanBapst It could actually be that the heading is initialized on a noisy mag sample; then, since the rover has less acceleration than a typical UAV, the heading takes more time to converge to the true value

File: src/lib/sensor_calibration/Magnetometer.cpp [hunk 3]
```diff
@@ -36,6 +36,7 @@
 #include "Utilities.hpp"
 
 #include <lib/parameters/param.h>
+#include <px4_platform_common/events.h>
```
Comment (dagar) [inline src/lib/sensor_calibration/Magnetometer.cpp:39]:
```suggestion
```

**Comment by dagar** [comment]
> @dagar @bresch I was helping @KonradRudin to figure out why the MAVROS rover tests are failing on his PR. Looks like it fails here as well. From what I could see the mean error between the ground truth yaw and the estimated yaw is too large. I plotted it and it looks like this: image
> 
> Any ideas?

@RomanBapst if you have the log on hand can you get an issue going? If we can easily grab a comparison log from before the change that couldn't hurt as well.

**Review by dagar — APPROVED** [review]

**Comment by PetervdPerk-NXP** [comment]
Any chance we can read out individual mag rotation before rotation correction so that we can which of the magnetometers producing wrong values?

Right now it's a pain to determine which mag is wrong when you get the magnetometers mismatch preflight check error.

**Pull request merged**